native/exe/ServiceControlLaunch/ServiceControlLaunch.cpp

```diff
@@ -173,60 +210,44 @@ int APIENTRY WinMain(HINSTANCE hInstance,
 	
 	if (createJavaVM==NULL)
 	{
-		errorMsg("Could not execute jvm.dll.\nPlease install Java Runtime Environment 1.4", "Java Missing");
+		errorMsg("Could not execute jvm.dll.\nPlease install Java Runtime Environment 1.7 or higher (Java 1.8 preferred)", "Java VM Creation Failed");
 		return FALSE;
 	}
 
-	HMODULE exeMod = GetModuleHandle(NULL);
-	LPTSTR appPath = new TCHAR[512];
-	GetModuleFileName(exeMod, appPath, 512);
-	size_t appLen = strlen(appPath);
-	if (appLen > 0)  // Shouldn't be 0 as the following would be an infinite loop
-	{
-		for (size_t i = appLen - 1; i > 0; i--)
-		{
-			if (appPath[i] == '\\')
-			{
-				appPath[i + 1] = 0;
-				break;
-			}
-		}
-	}
-
 	// Set the current working directory to be the folder the EXE is in.
-	chdir(appPath);
+	_chdir(appPath);
 
-	LPTSTR jarPath = new TCHAR[1024];
-	LPTSTR libraryPath = new TCHAR[512];
+	// Set up the JAVA VM
+	int const jarPathsize = 1024; // Define the size of jarPath
+	int const libraryPathsize = 512; // Define the size of libraryPath
+	LPTSTR jarPath = new TCHAR[jarPathsize];
+	LPTSTR libraryPath = new TCHAR[libraryPathsize];
 	
     JNIEnv *env;       /* pointer to native method interface */
 	JavaVMInitArgs vm_args;
 	JavaVMOption options[32];
 	vm_args.nOptions = 0;
-	strcpy(jarPath, "-Djava.class.path=");
-	strcat(jarPath, appPath);
+	strcpy_s(jarPath, jarPathsize, "-Djava.class.path=");

```

Comment:
While using separate constants isn't totally bad...putting them inline with the allocations and then using sizeof is much cleaner.

Reply:
I was running into issues with sizeof as jarPath isn't initialized until after strcpy_s

Reply:
The jarPath line and the libraryPath line should just be changed to:

char jarPath[1024];
char libraryPath[1024];

and then sizeof will work fine and you can remove the free statements at the end...man, this was such bad coding I did...ick. Or maybe there was some reason I was using dynamic instead of stack allocation...but you'll find out when you try to change it if you get some weird compiler errors. :)

Reply:
Done